File: scripts/update_tools_json.js

const fs = require('fs');
const path = require('path');

const toolsJsonPath = path.join(process.cwd(), 'src/constants/tools.json');
const toolsData = require(toolsJsonPath);

const newTools = [
  "audio-joiner",
  "video-watermark-remover",
  "audio-splitter",
  "background-remover",
  "video-background-remover",
  "youtube-summary-generator",
  "video-joiner",
  "image-to-icon",
  "developer-tools",
  "audio-normalizer",
  "audio-player",
  "faceless-video-generator",
  "video-resizer",
  "zoom-recorder",
  "video-to-audio",
  "generators",
  "pdf-protector",
  "instagram-reel-maker",
  "code-tools",
  "remove-duplicates",
  "images-to-pdf",
  "audio-spectrum",
  "course-video-maker",
  "video-sales-letter",
  "batch-image-processor",
  "pdf-reader",
  "audio-equalizer",
  "text-encoder",
  "ai-video-summarizer",
  "video-splitter",
  "video-rotator",
  "youtube-shorts-creator",
  "watermark-remover",
  "all-downloaders",
  "website-analyzer",
  "text-reverser",
  "line-sorter",
  "meme-video-generator",
  "case-converter",
  "pdf-to-images",
  "podcast-video-creator",
  "image-optimizer"
];

// Ensure 'others' category exists
if (!toolsData.categories.others) {
  toolsData.categories.others = {
    name: "Other Tools",
    slug: "other-tools",
    icon: "Grid", 
    description: "Various useful tools",
    tools: []
  };
}

// Flatten existing tools to check for duplicates
const existingIds = new Set();
Object.values(toolsData.categories).forEach(cat => {
  if (cat.tools) {
    cat.tools.forEach(t => existingIds.add(t.id));
  }
});

let addedCount = 0;

newTools.forEach(toolId => {
  if (existingIds.has(toolId)) {
    console.log(`Skipping ${toolId}, already exists (maybe in filtered list)`);
    return;
  }
  
  // Basic entry
  const toolName = toolId.split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
  const entry = {
    id: toolId,
    name: toolName,
    description: `${toolName} - Coming Soon`,
    route: `/${toolId}`,
    popular: false,
    category: "others"
  };

  toolsData.categories.others.tools.push(entry);
  existingIds.add(toolId);
  addedCount++;
});

// Update metadata
toolsData.metadata.totalTools += addedCount;
toolsData.metadata.totalCategories = Object.keys(toolsData.categories).length;

fs.writeFileSync(toolsJsonPath, JSON.stringify(toolsData, null, 2));
console.log(`Added ${addedCount} tools to tools.json`);

File: scripts/update_tools_json.py
import json
import os

base_path = '/Users/shaswatraj/Desktop/earn/30tools'
mapping_path = os.path.join(base_path, 'mapping_results.json')
tools_path = os.path.join(base_path, 'src/constants/tools.json')

with open(mapping_path, 'r') as f:
    mapping = json.load(f)

with open(tools_path, 'r') as f:
    original_tools_data = json.load(f)

def slug_to_title(slug):
    return ' '.join(word.capitalize() for word in slug.split('-'))

category_actions = {
    'image': 'compress, convert, and edit images',
    'pdf': 'manage PDF documents',
    'video': 'process video files',
    'text': 'transform and analyze text',
    'developer': 'format and convert code data',
    'downloaders': 'save social media content',
    'seo': 'optimize your website',
    'audio': 'edit audio files',
    'youtube': 'download YouTube content',
    'generators': 'generate random content',
    'utilities': 'perform helpful calculations',
    'others': 'perform online tasks'
}

# 1. Update existing tools with extra Slugs
tools_lookup = {}
for cat_id, category in original_tools_data['categories'].items():
    for tool in category['tools']:
        tools_lookup[tool['id']] = tool

for tool_id, extra in mapping['extraSlugsMapping'].items():
    if tool_id in tools_lookup:
        existing_extra = set(tools_lookup[tool_id].get('extraSlugs', []))
        existing_extra.update(extra)
        tools_lookup[tool_id]['extraSlugs'] = sorted(list(existing_extra))

# 2. Create new tools
for item in mapping['newTools']:
    master = item['master']
    extra = item['extraSlugs']
    category = item['category']
    name = slug_to_title(master)
    action = category_actions.get(category, 'perform online tasks')
    
    new_tool = {
        "id": master,
        "name": name,
        "description": f"Free {name} online. {name} allows you to {action} quickly and easily. 100% free, no signup required, and privacy-focused processing in your browser.",
        "route": f"/{master}",
        "extraSlugs": sorted(list(set(extra))),
        "category": category,
        "seoTitle": f"{name} - Free Online {category.capitalize()} Tool | 30tools",
        "seoDescription": f"Use our free {name} to {action} instantly. No registration needed, high quality results, and secure browser-based processing. Try {master} now!",
        "keywords": [master] + [w for w in master.split('-') if len(w) > 3],
        "features": [
            "100% Free to use",
            "No registration or signup required",
            "Fast browser-based processing",
            "Privacy-focused - data stays on your device",
            "Works on all modern browsers"
        ],
        "faqs": [
            {
                "question": f"Is this {name} free?",
                "answer": f"Yes, our {name} is completely free to use with no hidden costs or limits."
            },
            {
                "question": "Is my data secure?",
                "answer": "Absolutely. All processing happens locally in your browser. We never store or transmit your data to any server."
            }
        ],
        "howTo": {
            "name": f"How to use {name}",
            "steps": [
                {
                    "name": "Input Data",
                    "text": "Provide your input or upload your file to the tool.",
                    "url": f"/{master}#step1",
                    "position": 1
                },
                {
                    "name": "Process",
                    "text": "Click the process button to start the tool.",
                    "url": f"/{master}#step2",
                    "position": 2
                },
                {
                    "name": "Save Results",
                    "text": "Download or copy your results instantly.",
                    "url": f"/{master}#step3",
                    "position": 3
                }
            ]
        }
    }
    
    # Add to original data
    if category in original_tools_data['categories']:
        original_tools_data['categories'][category]['tools'].append(new_tool)
    else:
        # others is a fallback
        original_tools_data['categories']['others']['tools'].append(new_tool)

with open(os.path.join(base_path, 'src/constants/tools.json'), 'w') as f:
    json.dump(original_tools_data, f, indent=2)

print(f"Updated tools.json with {len(mapping['newTools'])} new tools and {len(mapping['extraSlugsMapping'])} slug updates.")
